File: evergreen/Scenarios.cpp
/* 
 * File:   Scenarios.cpp
 * Author: Weiming Hu <weiming@psu.edu>
 * 
 * Created on March 13, 2020, 1:08 PM
 */

#include "Scenarios.h"
#include "Functions.h"

#include <cmath>

using namespace std;

/******************************************************************************
 *                                    Scenario                                *
 ******************************************************************************/

void
Scenario::set(ssc_data_t & data_container) const {

    for (const auto & e : * this) {
        ssc_data_set_number(data_container, e.first.c_str(), e.second);
    }

    return;
}


void
Scenario::print(ostream & os) const {
    os << "{ ";
    for (const auto & e : *this) {
        os << e.first << ":" << e.second << " ";
    }
    os << "}";
    return;
}

ostream &
operator<<(ostream & os, const Scenario & scenario) {
    scenario.print(os);
    return os;
}


/******************************************************************************
 *                                   Scenarios                                *
 ******************************************************************************/


size_t
Scenarios::totalScenarios() const {

    if (size() == 0) return 0;

    size_t total = 1;
    for (const auto & e : *this) total *= e.second.size();
    return total;
}

void
Scenarios::set(ssc_data_t & data_container, size_t index) const {
    Scenario scenario;
    get(scenario, index);
    scenario.set(data_container);
    return;
}

void
Scenarios::get(Scenario & scenario, size_t index) const {

    // Clear map
    scenario.clear();

    // Initialize values
    size_t accum_len = 1;
    size_t current_index;

    for (const auto & e : * this) {

        // Calculate the current index for this key based on the global index
        if (accum_len == 1) {
            accum_len = e.second.size();
            current_index = fmod(index, accum_len);
        } else {
            current_index = fmod(index / accum_len, e.second.size());
            accum_len *= e.second.size();
        }

        // Insert pair
        scenario.insert(make_pair(e.first, e.second.at(current_index)));
    }

    return;
}

void
Scenarios::print(ostream & os) const {
    os << "Scenarios:" << endl;
    for (const auto & e : *this) {
        os << e.first << ":" << Functions::format(e.second, ",") << endl;
    }
    return;
}

void
Scenarios::printMore(ostream & os) const {
    os << "A list of available scenarios:" << endl;
    
    size_t total = totalScenarios();
    for (size_t i = 0 ; i < total; ++i) {
        Scenario scenario;
        get(scenario, i);
        os << "Scenario #" << i << "/" << total << ": " << scenario << endl;
    }
    
    return;
}

ostream &
operator<<(ostream & os, const Scenarios & scenarios) {
    scenarios.print(os);
    return os;
}

File: Scenarios.py
# "`-''-/").___..--''"`-._
#  (`6_ 6  )   `-.  (     ).`-.__.`)   WE ARE ...
#  (_Y_.)'  ._   )  `._ `. ``-..-'    PENN STATE!
#    _ ..`--'_..-_/  /--'_.' ,'
#  (il),-''  (li),'  ((!.-'
#
# Author:
#
#     Weiming Hu <weiming@psu.edu>
#
# Geoinformatics and Earth Observation Laboratory (http://geolab.psu.edu)
# Department of Geography and Institute for CyberScience
# The Pennsylvania State University
#

import os
import math
import yaml


class Scenarios(dict):
    """
    Scenarios is the class that stores and define multiple scenarios that will be simulated through the renewable
    simulator. It is inherited from the python built-in class, dict.

    The dictionary values only accept tuples or lists.

    A single scenario will be defined by a certain combination of values from all keys. For example, scenario #1 is
    defined using the first value in all keys; scenario #2 is defined using the second value in the first key and
    the first value in all other keys.
    """

    def total_scenarios(self):
        """
        Counts the total number of scenarios that could be defined using the current object of the class Scenarios.
        The total number is calculated from the multiplication of the length of each dictionary entry.

        :return: the number of total scenarios
        """
        total = 1

        for key in self:
            if isinstance(self[key], (list, tuple)):
                total *= len(self[key])
            else:
                raise Exception('Scenarios only allow lists or tuples as values')
        return total

    def get_scenario(self, index):

        # Check out-of-bound issues
        total_scenarios = self.total_scenarios()
        if index >= total_scenarios:
            raise Exception("Index {} is out of bound. The toal number of scenarios is {}".format(index, total_scenarios))

        # Initialize variables
        accumulate_len = 1
        current_scenario = {}

        for key, value in self.items():

            # Calculate the current index for this configuration
            if accumulate_len == 1:
                accumulate_len = len(value)
                current_index = index % accumulate_len
            else:
                current_index = math.floor(index / accumulate_len) % len(value)
                accumulate_len *= len(value)

            current_scenario[key] = value[current_index]

        return current_scenario

    def write_scenarios(self, file, overwrite=True):

        # Check output file
        if os.path.exists(file):
            if overwrite:
                os.remove(file)
            else:
                raise Exception("Output file {} exists and overwriting is not permitted".format(file))

        # Initialize a dictionary
        scenarios_dict = {}

        # Convert scenarios to dictionary with names
        for scenario_index in range(self.total_scenarios()):

            # Get the current scenario and the name
            scenario_name = 'scenario_{:05d}'.format(scenario_index)
            scenarios_dict[scenario_name] = self.get_scenario(scenario_index)

        # Write the current scenario
        with open(file, 'w') as f:
            yaml.dump(scenarios_dict, f)

    def __str__(self):
        msg = '******************** Scenarios ********************'
        msg += '\nTotal scenarios: {}'.format(self.total_scenarios())
        for k, v in self.items():
            msg += '\n  - {}: {} scenario(s)'.format(k, len(v))
        msg += '\n**************** End of Scenarios *****************'
        return msg
